File: python/leetcodehard/split_array_largest_sum.py
"""
https://leetcode.com/problems/split-array-largest-sum/

status - completed
"""

from base_logger.logging_event import create_logger


class Solution:

    log = create_logger(__name__)

    def splitArray(self, nums, m):
        low, high = max(nums), sum(nums)

        while low < high:
            mid = (low + high) // 2
            if self.count_splits(nums, mid) <= m:
                high = mid
            else:
                low = mid + 1

        return low

    def count_splits(self, nums, max_sum):
        splits = 1
        current_sum = 0

        for num in nums:
            if current_sum + num > max_sum:
                splits += 1
                current_sum = num
            else:
                current_sum += num

        return splits


if __name__ == "__main__":
    test_input_1 = ([7, 2, 5, 10, 8], 2)
    output = Solution().splitArray(*test_input_1)
    Solution().log.info("Output is %s", output)

    test_input_2 = ([1, 2, 3, 4, 5], 2)
    output = Solution().splitArray(*test_input_2)
    Solution().log.info("Output is %s", output)

File: typescript/leetcodehard/splitArrayLargestSum.ts
/*
https://leetcode.com/problems/split-array-largest-sum/

status - completed
*/

import { logp } from "../utils/logger";

function splitArray(nums: number[], m: number): number {
  let low = Math.max(...nums);
  let high = nums.reduce((sum, n) => sum + n, 0);

  function countGroups(maxSum: number): number {
    let groups = 1;
    let currentSum = 0;

    for (const num of nums) {
      if (currentSum + num > maxSum) {
        groups++;
        currentSum = num;
      } else {
        currentSum += num;
      }
    }

    return groups;
  }

  while (low < high) {
    const mid = low + Math.floor((high - low) / 2);
    if (countGroups(mid) > m) {
      low = mid + 1;
    } else {
      high = mid;
    }
  }

  return low;
}

logp(`input nums=[7,2,5,10,8] m=2 expected 18 actual ${splitArray([7, 2, 5, 10, 8], 2)}`);
logp(`input nums=[1,2,3,4,5] m=2 expected 9 actual ${splitArray([1, 2, 3, 4, 5], 2)}`);
logp(`input nums=[1,4,4] m=3 expected 4 actual ${splitArray([1, 4, 4], 3)}`);
